_posts/2022-12-12-deltalake-spark-testing-learnings.md

```diff
@@ -0,0 +1,232 @@
+
+CMD export PYSPARK_SUBMIT_ARGS='--packages io.delta:delta-core_2.12:2.2.0 pyspark-shell' \
+    pytest path/to/tests/**/*
+```
+
+If you are wondering why theres no Apache setup steps above, then remember that we are will be installing pyspark via pip instead which will suffice for local development activities. 
+
+Most of the other config is stock standard, with the exception of the `PYSPARK_SUBMIT_ARGS`. Its one of those things you need to set to run these tests as a workaround certain known issues with running locally. At the end of the day, these are just arguments that end up being supplied to `spark submit` which is a script that Spark uses to launch apps on a cluster. 
+
+Why you might still ask? well, the purpose behind each argument is as follows:
+
+1. io.delta:delta-core_2.12:2.1.0 - defines the maven artifact dependency of delta-core_2.21 for the version 2.1.0. You must match the version with precisely what you defined as the `delta-spark` pip dependency. If you are not sure about the artifact name, here is a reference: https://mvnrepository.com/artifact/io.delta/delta-core. The artifact name will be different but the version is whats important.
+
+2. pyspark-shell specifies to use the pyspark java class when running spark jobs locally. 
+
+
+*docker-compose.yml*
+```
+version: "3.9"
+
+services:
+  test:
+    build:
+      context: .
+      dockerfile: Dockerfile
+    container_name: test
+    volumes:
+      - .:/src
+```
+
+Then we just need to run docker-compose up to re-run the tests whenever we make changes or create new ones locally. Of course, the volume mount in the compose file will depend on your project structure.
+
+
+### Create Configure Spark Session In Test Hook
+
+At this point, we can focus on what the code requires. We will need a before hook that configures our local spark session so our tests can use it. This implication here is that any delta tables that created will be placed our local file system, but thats the advantage of it I guess when it comes to testability.
+
+In your before hook for all tests, you will need to create a local spark session object (see code below). What's important is that we use `configure_spark_with_delta_pip` to ensure that the Delta Lake Maven JARs are installed and is hence one of the few reasons why we specified the `delta-spark` dependency to begin with. 
+
+```
+from pyspark.sql.session import SparkSession
+from delta import configure_spark_with_delta_pip
+import pytest
+
+@pytest.fixture(scope="session")
+def spark():
+    builder = SparkSession.builder.appName("local-spark-session").config("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension").config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog")
+
+    return configure_spark_with_delta_pip(builder).getOrCreate()
+```
+
+See the docs local [delta lake setup](https://docs.delta.io/latest/quick-start.html#python) if you wish to learn more.
+
+### Writing Test
```

Comment:
```suggestion
### Writing Tests
```
or

```suggestion
### Writing Our First Test
```

minor wording suggestion for readability

Reply:
going to go for `Writing Our First Test`